app/services/proxy_deployment_service.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+class ProxyDeploymentService
+  delegate :apicast_configuration_driven,
+           :deployable?,
+           :service_mesh_integration?,
+           :provider,
+           :proxy_configs, to: :@proxy
+
+  def self.call(*args)
+    new(*args).call
+  end
+
+  def initialize(proxy, stage = :staging)
+    @proxy = proxy
+    @stage = stage
+  end
+
+  def call
+    case @stage
+    when :staging, :sandbox
+      deploy
+    when :production
+      deploy_production
+    else
+      raise "nope" # ??
```

Comment:
Yes :) Use a tagged error and rescue it where needed:

- controller
- sidekiq 😱 Yes because we do async deploy https://github.com/3scale/porta/blob/3968319a9fa477de9997ad7860bbc8ab5268d465/app/lib/logic/rolling_updates.rb#L160-L164